### PR TITLE
Node-Red multi-arch Node v10 Docker images [amd64, arm32v6, arm32v7, arm64v8]

### PR DESCRIPTION
**Pull request by RaymondMouthaan**
This PR adds multi-architecture (amd64, arm32v6, arm32v7 and arm64v8) support for node-red docker images  and is manifest listed. The base image is node v10.

### CODE REVIEW

**Comment by knolleary** [comment]
Hi, thanks for this. I won't be able to review properly for a day or two, but rest assured we're very interested in incorporating your work in the official images.

Can you expand in the description a bit more? Which of the tagged images we produce does this apply to? Are there any other functional changes to the images other than the multiarch support?

**Comment by RaymondMouthaan** [comment]
Hi @knolleary, this PR is where we discussed about several months ago, but since I saw many discussions lately regarding this topic, so I though to do do the PR once again :)

The out-come of this PR can be found [here](https://cloud.docker.com/repository/docker/raymondmm/node-red/general) and basically does the following:

| Image Tag                | Architectures                    | OS           | Node | Description                                   |
|--------------------------|----------------------------------|--------------|------|-----------------------------------------------|
| 0.20.5-10-alpine-amd64   | amd64                            | Alpine Linux | v10  | Node-red image for amd64, based on node v10   |
| 0.20.5-10-alpine-arm32v6 | arm32v6                          | Alpine Linux | v10  | Node-red image for amd32v6, based on node v10 |
| 0.20.5-10-slim-arm32v7   | arm32v7                          | Debian Slim  | v10  | Node-red image for arm32v7, based on node v10 |
| 0.20.5-10-alpine-arm64v8 | arm64v8                          | Alpine Linux | v10  | Node-red image for arm64v8, based on node v10 |
| latest                   | amd64, arm32v6, arm32v7, arm64v8 |              | v10  | Manifest list for all images above            |

_note: we might consider replacing `0.20.5-10-slim-arm32v7` with Alpine Linux from [here](https://hub.docker.com/r/arm32v7/alpine/tags), that way all the images are Alpine based with smallest size._

- creation of 4 node-red images based on node v10, for arm32v6, arm32v7, arm64v8 and amd64.
- it manifest lists those images so there is no need to explicit apply a tag, docker auto detects which image it requires.
- it uses travis 
- as far as I can tell, there are no functional changes, only one thing worth mentioning is [this](https://github.com/node-red/node-red-docker/issues/60#issuecomment-497988888) and [this](https://github.com/node-red/node-red-docker/issues/60#issuecomment-498008522), which are cover by [Fix [warn] Cannot find Pi RPi.GPIO python library](https://github.com/RaymondMouthaan/node-red-docker/issues/11)


I am using these images on a docker swarm for several months now, without any issues and I'll try to keep them up to date with the latest release of node-red, but I prefer to have it part of the official node-red docker repo instead of my own.

**Comment by ernestgwilsonii** [comment]
Does this add the official arm32v7/node:
https://hub.docker.com/r/arm32v7/node

Instead of the now deprecated hypriot/rpi-node:
https://hub.docker.com/r/hypriot/rpi-node/

**Comment by RaymondMouthaan** [comment]
Yes it does pull the official arm images, look close and you'll find it ;)

**Comment by ernestgwilsonii** [comment]
Excellent! Nice work! Glad this is getting updated in time for version 1.0.0

File: .travis.yml [hunk 1]
```diff
@@ -1,49 +1,76 @@
...
-  - docker
+  - 'docker'
+
+language:
+  - 'bash'
+
+env:
+  global:
+    - NODE_RED_VERSION=x.y.z
+    - TARGET=raymondmm/node-red
```
Comment (dceejay) [inline .travis.yml]: just checking - this is just a temp dir for build ?

File: .travis.yml [hunk 2]
```diff
@@ -1,49 +1,76 @@
...
+    - QEMU_VERSION=v3.0.0
+  matrix:
+    - DOCKER_FILE=Dockerfile.alpine-tmpl BUILD_FROM=amd64/node NODE_VERSION=10 OS=alpine QEMU_ARCH=x86_64 ARCH=amd64
+    - DOCKER_FILE=Dockerfile.alpine-tmpl BUILD_FROM=arm32v6/node NODE_VERSION=10 OS=alpine QEMU_ARCH=arm ARCH=arm32v6
+    - DOCKER_FILE=Dockerfile.slim-tmpl BUILD_FROM=arm32v7/node NODE_VERSION=10 OS=slim QEMU_ARCH=arm ARCH=arm32v7
```
Comment (dceejay) [inline .travis.yml]: Should we have a full fat version with all the dev/build tools in ? python node-gyp build-essential etc

File: .travis.yml [hunk 3]
```diff
@@ -1,49 +1,76 @@
...
+language:
+  - 'bash'
+
+env:
+  global:
+    - NODE_RED_VERSION=x.y.z
+    - TARGET=raymondmm/node-red
+    - QEMU_VERSION=v3.0.0
+  matrix:
+    - DOCKER_FILE=Dockerfile.alpine-tmpl BUILD_FROM=amd64/node NODE_VERSION=10 OS=alpine QEMU_ARCH=x86_64 ARCH=amd64
```
Comment (dceejay) [inline .travis.yml]: Should we just do v10 ? or should we include v12 (for testing only ?)

File: README.md [hunk 4]
```diff
@@ -1,35 +1,54 @@
 # Node-RED-Docker
 
-[![Greenkeeper badge](https://badges.greenkeeper.io/node-red/node-red-docker.svg)](https://greenkeeper.io/)
+[![Greenkeeper badge](https://badges.greenkeeper.io/RaymondMouthaan/node-red-docker.svg)](https://greenkeeper.io/)
```
Comment (dceejay) [inline README.md]: should be node-red path etc

File: README.md [hunk 5]
```diff
@@ -1,35 +1,54 @@
...
+Running that command should give a terminal window with a running instance of Node-RED.
 
         Welcome to Node-RED
+
         ===================
-        8 Apr 12:13:44 - [info] Node-RED version: v0.14.5
-        8 Apr 12:13:44 - [info] Node.js  version: v4.4.7
-        .... etc
+
+        6 May 00:36:30 - [info] Node-RED version: v0.18.4
```
Comment (dceejay) [inline README.md]: Should at least match latest versions for now :-)

File: README.md [hunk 6]
```diff
@@ -42,45 +61,183 @@ If we are happy with what we see we can detach the terminal with `Ctrl-p``Ctrl-q
...
 
-Using Alpine Linux reduces the built image size (~100MB vs ~700MB) but removes
+The Node-RED images are based on the [official Node JS v8 and Node JS v6 images](https://hub.docker.com/_/node/), which are based on Alpine Linux or Debian Linux (slim).
+Using Alpine Linux or Debian Linux (slim) reduces the built image size (~100MB vs ~700MB), but removes
 standard dependencies that are required for native module compilation. If you
-want to add modules with native dependencies, use the standard image or extend
-the slim image with the missing packages.
+want to add dependencies with native dependencies, extend the Node-RED image with the missing packages on running containers or build new images.
+
+### Node.JS v8 based images
```
Comment (dceejay) [inline README.md]: Should match the versions being built

File: README.md [hunk 7]
```diff
@@ -96,7 +253,7 @@ Docker build process, the dependencies are installed under `/usr/src/node-red`.
 The main sections to modify are
 
     "dependencies": {
-        "node-red": "0.14.x",           <-- set the version of Node-RED here
+        "node-red": "0.18.x",           <-- set the version of Node-RED here
```
Comment (dceejay) [inline README.md]: likewise should match latest available

File: package.json [hunk 8]
```diff
@@ -21,10 +21,13 @@
...
             "name": "James Thomas"
+        },
+        {
+            "name": "Raymond Mouthaan"
         }
     ],
     "dependencies": {
-        "node-red": "0.18.4",
+        "node-red": "0.20.5",
         "node-red-node-msgpack": "*",
```
Comment (dceejay) [inline package.json]: Extra nodes not really needed - were here to sort of match the Pi pre-install - but are not core nodes.
There is also an engines statement in this file that should be updated/fixed to at least 8.x

**Comment by RaymondMouthaan** [comment]
@dceejay, thanks for your feedback :)

It has been a while, but meanwhile I have been working on a new PR regarding Node-RED Docker images.

In a nutshell the new PR will cover the following:
- Multi-arch support for arm32v6, arm32v7, arm64v8, amd64
- Node 10
- Python 2 and Python 3
- Node-RED build-in RPI-GPIO support
- Keep the images as small as possible (preferable Alpine based)

Currently all images and are manifest lists are tested on best effort and I am updating the README, as well taking into account the comments you made above. 

[This](https://github.com/RaymondMouthaan/node-red-docker/tree/dev-alpine-buster-slim) branch refers to Docker repo raymondmm/node-red, but once it is finished, I'll create a new branch and point to the official Docker Node-RED repo and pull a new request and finally merge it into the node-red-docker repo :) 

Cheers,
Ray

File: .travis.yml (hunk 1 above)
Comment (RaymondMouthaan) [reply .travis.yml]: NODE_RED_VERSION is a environment variable which gets filled by update_version.sh

TARGET is a environment variable which holds the target Docker repo, in this case raymondmm/node-red, but will change to nodered/node-red-docker or even better to nodered/node-red (without "-docker"), but this means a new repo on Docker hub. In my humble opinion `nodered/node-red-docker` is a pleonasm and should be left as is, while all new stuff should go into `nodered/node-red`.

File: .travis.yml (hunk 2 above)
Comment (RaymondMouthaan) [reply .travis.yml]: If you can list all tools which might come in handy for a full fat version, i can adapt these in:
- 0.20.7-alpine-amd64-devtools
- 0.20.7-alpine-arm32v6-devtools
- 0.20.7-buster-slim-arm32v7-devtools 
- 0.20.7-alpine-arm64v8-devtools 

And manifest list them in:
- 0.20.7-devtools
- latest-devtools

File: .travis.yml (hunk 3 above)
Comment (RaymondMouthaan) [reply .travis.yml]: I'll like to purpose to merge the new PR first and build all images according the new way. If all works fine, we can easily adopt Node v12.

File: README.md (hunk 4 above)
Comment (RaymondMouthaan) [reply README.md]: Taking care of this in the new coming PR.

File: README.md (hunk 5 above)
Comment (RaymondMouthaan) [reply README.md]: Taking care of this in the new coming PR.

File: README.md (hunk 6 above)
Comment (RaymondMouthaan) [reply README.md]: Taking care of this in the new coming PR.

File: README.md (hunk 7 above)
Comment (RaymondMouthaan) [reply README.md]: Taking care of this in the new coming PR.

File: package.json (hunk 8 above)
Comment (RaymondMouthaan) [reply package.json]: The package.json in the new coming PR looks like this:

```
...
    "dependencies": {
        "node-red": "0.20.7"
    },
    "engines": {
        "node": "8.*.*"
    }
...
```

Would that be sufficient?

**Comment by RaymondMouthaan** [comment]
I am closing this PR, since a new PR is coming ...

**Pull request closed without merging**